third_party/blink/renderer/core/resource_pool_limiter/resource_pool_limiter.cc

```diff
@@ -0,0 +1,112 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/third_party/blink/renderer/core/resource_pool_limiter/resource_pool_limiter.h"
+
+#include "base/notreached.h"
+#include "third_party/blink/renderer/core/execution_context/security_context.h"
+#include "third_party/blink/renderer/core/frame/local_dom_window.h"
+#include "third_party/blink/renderer/platform/weborigin/security_origin.h"
+#include "third_party/blink/renderer/platform/wtf/std_lib_extras.h"
+#include "third_party/blink/renderer/platform/wtf/text/string_builder.h"
+
+namespace blink {
+
+using ResourceType = ResourcePoolLimiter::ResourceType;
+
+namespace {
+
+const SecurityOrigin* GetTopFrameOrContextSecurityOrigin(
+    ExecutionContext* context) {
+  DCHECK(context);
+  if (auto* window = DynamicTo<LocalDOMWindow>(context)) {
+    auto* frame = window->GetFrame();
+    if (!frame)
+      frame = window->GetDisconnectedFrame();
+    if (frame) {
+      if (auto* top_frame_security_context =
+              frame->Top()->GetSecurityContext()) {
+        return top_frame_security_context->GetSecurityOrigin()
+            ->GetOriginOrPrecursorOriginIfOpaque();
+      }
+    }
+  }
+  return context->GetSecurityOrigin()->GetOriginOrPrecursorOriginIfOpaque();
+}
+
+String GetResourceIdInUse(const SecurityOrigin* origin,
+                          ResourceType resource_type) {
+  DCHECK(origin);
+  StringBuilder string_builder;
+  string_builder.AppendNumber(static_cast<int>(resource_type));
+  String origin_id = origin->RegistrableDomain();
+  if (origin_id.IsEmpty())
+    origin_id = origin->Host();
+  string_builder.Append(origin_id);
+  return string_builder.ToString();
+}
+
+int GetResourceLimit(ResourceType resource_type) {
+  switch (resource_type) {
+    case ResourceType::kWebSocket:
+      return 10;
```

Comment:
Should this be hard-coded? Would it make sense to make it a feature param, for example, and control it via Griffin?